src/api/api.js

```diff
@@ -1,16 +1,50 @@
-import { getInstance as getD2Instance } from 'd2'
-import { pageSize } from '../constants/development'
-import { supportsUserLookupEndPoint } from '../utils/helpers.js'
+import { pageSize } from '../constants/development.js'
+import createRecipientSearchQuery from '../utils/createRecipientSearchQuery.js'
+
+const initialMessageConversationFields = [
+    'id',
+    'displayName',
+    'subject',
+    'messageType',
+    'lastSender[id, displayName]',
+    'assignee[id, displayName]',
+    'status',
+    'priority',
+    'lastUpdated',
+    'read',
+    'lastMessage',
+    'followUp',
+]
+
+const messageConversationFields = [
+    '*',
+    'assignee[id, displayName]',
+    'messages[*,sender[id,displayName]',
+    'attachments[id, name, contentLength]]',
+    'userMessages[user[id, displayName]]',
+]
 
-const initialMessageConversationFields =
-    'id, displayName, subject, messageType, lastSender[id, displayName], assignee[id, displayName], status, priority, lastUpdated, read, lastMessage, followUp'
+const order = 'lastMessage:desc'
 
-const messageConversationFields =
-    '*,assignee[id, displayName],messages[*,sender[id,displayName],attachments[id, name, contentLength]],userMessages[user[id, displayName]]'
+// The rest of the code in this module will expect that
+// engine has been set and can use it safely
+let engine = null
+export const setEngine = engineInstance => {
+    engine = engineInstance
+}
```

Comment:
Should we not use the pattern from: https://runtime.dhis2.nu/#/advanced/redux?

Reply:
I discussed the used approach with you before I got started. Here's [a link to our DMs in Slack](https://dhis2.slack.com/archives/DH3B2NDQW/p1602060766014400). Since this is in Dutch and not accessible to anyone else, I'll just provide a summary of what we discussed below:
- I basically suggested the current approach and the pattern from the runtime docs
- I explained the current approach would be less invasive, but more quirky
- I understood from you that you would prefer the least invasive approach. Possibly I misinterpreted what you meant.

Happy to refactor if you prefer but that'll take some time.

Reply:
Ah sorry. Yeah I remember now. I didn't really get the entire scope of what you were suggesting at that time. But would that pattern be that invasive though? It's basically just a slightly different store implementation right? Or am I overseeing sth that would take a lot of work.

Reply:
I think it's probably doable but it would have been a lot more productive if I implemented it like this from the get go. Every function in the api module would get the `engine` as an arg. So that would mean:
- Change every function in the api module to receive the engine. If we have any functions that already have three params, these will have to be rewritten to get an options object.
- Every action that calls an api module will need to pull the engine off the state to pass it to api function. Again, some functions may need to be rewritten to using an options object...
- In some places of the app, functions from the api module are being called directly from a component. In these cases I'll probably just import `useEngine` and pass the engine that way...

Reply:
> Every action that calls an api module will need to pull the engine off the state to pass it to api function. Again, some functions may need to be rewritten to using an options object...

You don't have to get the engine from the state though. The thunk will just get an extra argument which is the engine.

> In some places of the app, functions from the api module are being called directly from a component. In these cases I'll probably just import useEngine and pass the engine that way...

Yes that seems most appropriate to me as well.

> I think it's probably doable but it would have been a lot more productive if I implemented it like this from the get go.

Yeah I understand, sorry about that. My agreeing with the suggestion to go with the current approach in our chat was based on me also preferring the less invasive approach, but at the moment I wonder if this is actually that much more invasive. If it seems like too much of a change for this PR then maybe it would be a solution to split up the work? Say look at this PR in isolation, and then implement the canonical pattern afterwards for example. That is also assuming that you agree with the points made. Could also be that you don't of course.

Reason I'm suggesting it is also because I wonder whether the current pattern will be a 100% reliable. I know that es6 modules will always be singletons, but I'm not so sure about that for this app, since transpilation might affect module behaviour? And stateful modules are harder to reason about in my opinion. That's why I'm being a bit cautious here.

Reply:
Agreed, I think it's worth the work. Starting on it now.

Reply:
I've looked into this and ran into a very quirky issue. MUI v0.20 has an issue with propagating the react context onto portals. As a result, components in a Portal can't access the DataContext. And since `@dhis2/app-runtime` does not expose the `DataContent`, we can't use the same trick we did for the redux context (see `src/components/Common/DialogWithReduxState.js`).

We discussed the option of adding a change to `@dhis2/app-runtime` 
 so that it does expose the `DataContext`, but it was decided that wasn't a good idea.

So in the end we are going with the "stateful api module" approach as we already had.